File: practise/daliy practise/Scientific_Computing_With_Machine/OpenMP/Do_for.f90
integer function work(val)
  implicit none
  integer val
  write (*,*) "worker ", val , "doing\n"
  work = val * 2
end function work

subroutine work1(val)
  implicit none
  integer val
  write (*,*) "worker ", val , "doing\n"
end subroutine work1



program dofor
  use omp_lib
  implicit none
  integer, external :: work
  integer n, i, return_val, k, count

  n = 11
  !$omp parallel private(k)
  !$omp do
  do i=1,n
     !return_val = work(i)
     call work1(i)
     k = omp_get_thread_num()
     print *, "do for worker k: ", k
     !print *, "work ", i
  end do
  !$omp end do
  k = omp_get_thread_num()
  count = omp_get_num_threads()
  write (*,*) "worker :", k , "out of :", count
  !$omp end parallel

end program dofor

File: practise/daliy practise/Scientific_Computing_With_Machine/OpenMP/do_for.c
#include<stdio.h>
#include<omp.h>
#include<unistd.h>

void work(int id, int i){
    printf("worker %d doing, i: %d\n", id, i);
    return;
}

int main(int argc, char* argv[]){
    //int i = 0, n = 10;
    int N = 14;
#pragma omp parallel if(N > 13)
    {
        int thread_id = omp_get_thread_num();
        work(thread_id, -1);
        int i = 0, n = 10;
#pragma omp for schedule(dynamic, 2)
        for(i = 0; i < n; i++){
            sleep(i + 1);
            int thread_id = omp_get_thread_num();
            work(thread_id, i);
            //sleep(i);
        }
        thread_id = omp_get_thread_num();
        work(thread_id, -1);
    }
    return 0;
}